### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by leosimas**
First version of this app

### CODE REVIEW

File: app/src/main/java/com/leosimas/moviesapp/ui/adapter/InfiniteRecyclerViewAdapter.java [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+    }
+
+    public List<T> getList() {
+        return list;
+    }
+
+    protected abstract K onCreateItemViewHolder(@NonNull ViewGroup parent);
+    protected abstract void onBindItemViewHolder(@NonNull K holder, int position);
+
+    private class TryAgainViewHolder extends RecyclerView.ViewHolder {
```
Comment (arctouch-danilocarvalho) [inline app/src/main/java/com/leosimas/moviesapp/ui/adapter/InfiniteRecyclerViewAdapter.java]: How do you feel about making these ViewHolder classes static? Would there be any benefits or drawbacks in doing that?

File: app/src/main/java/com/leosimas/moviesapp/ui/view/MovieView.java [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+import com.leosimas.moviesapp.util.FormatUtils;
+
+import org.androidannotations.annotations.Click;
+import org.androidannotations.annotations.EViewGroup;
+import org.androidannotations.annotations.ViewById;
+
+import java.util.List;
+
+@EViewGroup(R.layout.view_movie)
+public class MovieView extends RelativeLayout {
```
Comment (arctouch-danilocarvalho) [inline app/src/main/java/com/leosimas/moviesapp/ui/view/MovieView.java]: Good job creating your own Custom Views. Not a lot of people do that and I think it's a great way to encapsulate some reusable view logic into one place.

File: app/src/main/java/com/leosimas/moviesapp/util/AppUtils.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
+package com.leosimas.moviesapp.util;
+
+import android.content.Context;
+import android.support.annotation.StringRes;
+import android.view.inputmethod.InputMethodManager;
+import android.widget.EditText;
+import android.widget.Toast;
+
+public class AppUtils {
```
Comment (arctouch-danilocarvalho) [inline app/src/main/java/com/leosimas/moviesapp/util/AppUtils.java]: Do you have any experience with the Kotlin language? It helps replace a lot of these Util classes with extensions methods. Just one of the great features of the language.

File: app/src/main/res/layout/view_movie.xml [hunk 4]
```diff
@@ -0,0 +1,60 @@
+<?xml version="1.0" encoding="utf-8"?>
+<RelativeLayout xmlns:android="http://schemas.android.com/apk/res/android"
```
Comment (arctouch-danilocarvalho) [inline app/src/main/res/layout/view_movie.xml]: Have you ever used ConstraintLayout?

File: app/src/main/res/layout/view_movie.xml [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+    android:id="@+id/rootView"
+    android:layout_width="match_parent"
+    android:layout_height="wrap_content"
+    android:padding="@dimen/margin_medium"
+    android:background="?android:attr/selectableItemBackground">
+
+    <ImageView
+        android:id="@+id/imagePoster"
+        android:layout_width="100dp"
+        android:layout_height="150dp"
```
Comment (arctouch-danilocarvalho) [inline app/src/main/res/layout/view_movie.xml]: You used Dimen resources for the padding on line 7. Why not use it for the height and width as well? 😄

File: app/src/androidTest/java/com/leosimas/moviesapp/MoviesAPITest.java [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+        final CountDownLatch signal = new CountDownLatch(1);
+
+        final int page = 1;
+
+        CallAdapter<MoviePage> callAdapter = moviesAPI.upcoming(page);
+        callAdapter.enqueue(new MoviesAPI.Callback<MoviePage>() {
+            @Override
+            public void onSuccess(MoviePage result) {
+                assertEquals(page, result.getPage());
+                assertTrue(result.getTotalPages() > 0);
```
[the remark this answers is not in the document]
Comment (leosimas) [reply app/src/androidTest/java/com/leosimas/moviesapp/MoviesAPITest.java]: The objective of **MoviesAPITest** is exactally to test these requests. I used it as I was creating the services and beans classes.

File: README.md [hunk 7]
```diff
@@ -1,2 +1,20 @@
...
+- production
+
+## Third-party libraries
+This project is using some real nice frameworks and libraries listed below:
+- Retrofit - https://square.github.io/retrofit/
+    - responsable for the HTTP requests
+    - it also uses GSON for JSON convertion - https://github.com/google/gson
+- AndroidAnnotations - http://androidannotations.org/
+    - a framework that helps a lot with some code generation, object injection and make the code more simple to maintain
+- Parceler - http://parceler.org/
```
[the remark this answers is not in the document]
Comment (leosimas) [reply README.md]: I've been using AndroidAnnotations since 2012 and it really got better. Most of the Android native apps I've worked went to production using it. In the last years, I've been using with it in some bank apps.
In the past there were some cases where AndroidAnnotations could not help but it changed to support new features. In those cases we could use "pure Android" instead. I remember one time that I needed to use Generics and AndroidAnnotations did not supported it back then.
I've read something about it in the past but never tried.

File: app/src/main/java/com/leosimas/moviesapp/service/CallAdapter.java [hunk 8]
```diff
@@ -0,0 +1,79 @@
+package com.leosimas.moviesapp.service;
+
+import android.support.annotation.NonNull;
+import android.util.Log;
+
+import retrofit2.Call;
+import retrofit2.Callback;
+import retrofit2.Response;
+
+public class CallAdapter<T> {
```
[the remark this answers is not in the document]
Comment (leosimas) [reply app/src/main/java/com/leosimas/moviesapp/service/CallAdapter.java]: The idea is to handle the request response test here so all the callers don't need to test it. Check if successful and deliver only the body, movies list or any other result type.

File: app/src/main/java/com/leosimas/moviesapp/ui/MainActivity.java [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+                    buttonSearch();
+                    return true;
+                }
+                return false;
+            }
+        });
+
+    }
+
+    private void requestPage(final int page, final InfiniteRecyclerView.PageLoaderListener pageLoaderListener) {
```
[the remark this answers is not in the document]
Comment (leosimas) [reply app/src/main/java/com/leosimas/moviesapp/ui/MainActivity.java]: Something is not right... I've made some tests again and it's not working as you described.
The **MainActivity_** has already the **android:configChanges="orientation"** so it handles orientation changes instead of restarting it. The request it not been made again and the list position is still the same.

File: app/src/main/java/com/leosimas/moviesapp/ui/view/MovieView.java (hunk 2 above)
Comment (leosimas) [reply app/src/main/java/com/leosimas/moviesapp/ui/view/MovieView.java]: AndroidAnnotations makes it really easy :D

File: app/src/main/java/com/leosimas/moviesapp/util/AppUtils.java (hunk 3 above)
Comment (leosimas) [reply app/src/main/java/com/leosimas/moviesapp/util/AppUtils.java]: I've never worked with it on a project but I used some of its features during some Kotlin lessons.
It really reminded me of some Swift features.

File: app/src/main/res/layout/view_movie.xml (hunk 4 above)
Comment (leosimas) [reply app/src/main/res/layout/view_movie.xml]: I don't think so. I have to use constraints in iOS but I really don't like it.
Sometimes if feels too much work when creating simple layouts and it make doing changes harder.

File: app/src/main/res/layout/view_movie.xml (hunk 5 above)
Comment (leosimas) [reply app/src/main/res/layout/view_movie.xml]: I'm always reusing padding and margins values. But in this case, the imageview dimension was not going to be reused and I didn't created any layout modifiers for it.

File: app/src/main/java/com/leosimas/moviesapp/ui/MainActivity.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (arctouch-danilocarvalho) [reply app/src/main/java/com/leosimas/moviesapp/ui/MainActivity.java]: 🤔 weird. Let me try again to make sure.

File: README.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (arctouch-danilocarvalho) [reply README.md]: Cool. I'll try it some time. Thanks.

File: app/src/main/java/com/leosimas/moviesapp/ui/MainActivity.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (arctouch-danilocarvalho) [reply app/src/main/java/com/leosimas/moviesapp/ui/MainActivity.java]: I've just tested it and I had to add `screenSize` to the `android:configChanges` to get it to avoid restarting the activity. Maybe it works differently for different Android versions or phone manufacturers.

The documentation recommends using both `orientation` and `screenSize`: https://developer.android.com/guide/topics/resources/runtime-changes.

A better solution for handling orientation changes could be moving the state of the screen to another layer such as a ViewModel. You could use Android Architecture Components for that: https://developer.android.com/topic/libraries/architecture/

> ViewModel Stores UI-related data that isn't destroyed on app rotations.

File: app/src/main/java/com/leosimas/moviesapp/ui/adapter/InfiniteRecyclerViewAdapter.java (hunk 1 above)
Comment (leosimas) [reply app/src/main/java/com/leosimas/moviesapp/ui/adapter/InfiniteRecyclerViewAdapter.java]: I'm not sure if there are performance differences in this case. But I just remember that changing it to static it will not be possible to access members from the outter class directly from within the inner classes which it's not needed in this case.

File: app/src/main/java/com/leosimas/moviesapp/ui/MainActivity.java [hunk 10]
```diff
@@ -0,0 +1,229 @@
...
+        if (searchMode) {
+            callMovies = moviesAPI.search(lastSearchQuery, page);
+        } else {
+            callMovies = moviesAPI.upcoming(page);
+        }
+        callMovies.enqueue(new MoviesAPI.Callback<MoviePage>() {
```
[the remark this answers is not in the document]
Comment (leosimas) [reply app/src/main/java/com/leosimas/moviesapp/ui/MainActivity.java]: I could create another layer/class to handle the pages numbers and which request method should be called but the activity would still need to know which mode its showing.

File: app/src/main/java/com/leosimas/moviesapp/ui/MainActivity.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (leosimas) [reply app/src/main/java/com/leosimas/moviesapp/ui/MainActivity.java]: That's right!

File: app/src/main/java/com/leosimas/moviesapp/ui/MainActivity.java [hunk 11]
```diff
@@ -0,0 +1,229 @@
...
+        MenuItem searchItem = menu.getItem(0);
+        searchItem.setVisible( !searchMode );
+        return super.onPrepareOptionsMenu(menu);
+    }
+
+    private void updateMode() {
+        callMovies.cancel();
+
+        List<Movie> list;
+        if (searchMode) {
```
[the remark this answers is not in the document]
Comment (leosimas) [reply app/src/main/java/com/leosimas/moviesapp/ui/MainActivity.java]: It's the same problem about the missing "screenSize" in the android:configChanges.

File: app/src/androidTest/java/com/leosimas/moviesapp/MoviesAPITest.java [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.fail;
+
+@RunWith(AndroidJUnit4.class)
+public class MoviesAPITest {
```
[the remark this answers is not in the document]
Comment (leosimas) [reply app/src/androidTest/java/com/leosimas/moviesapp/MoviesAPITest.java]: I added the AndroidAnnotations after testing these requests so I missed that.
I should change the MoviesAPI member initialization and use the generated MoviesAPI_ class to let AndroidAnnotations handle the annotated members inside MoviesAPI.
